In [15]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

# Load Data

In [33]:
book = pd.read_csv('books_data/books.csv', encoding='latin-1', sep=';', 
                    error_bad_lines=False)
rating = pd.read_csv('books_data/ratings.csv', encoding='latin-1', sep=';', 
                    error_bad_lines=False)
user = pd.read_csv('books_data/users.csv', encoding='latin-1', sep=';', 
                    error_bad_lines=False)

Skipping line 6452: expected 8 fields, saw 9
Skipping line 43667: expected 8 fields, saw 10
Skipping line 51751: expected 8 fields, saw 9

Skipping line 92038: expected 8 fields, saw 9
Skipping line 104319: expected 8 fields, saw 9
Skipping line 121768: expected 8 fields, saw 9

Skipping line 144058: expected 8 fields, saw 9
Skipping line 150789: expected 8 fields, saw 9
Skipping line 157128: expected 8 fields, saw 9
Skipping line 180189: expected 8 fields, saw 9
Skipping line 185738: expected 8 fields, saw 9

Skipping line 209388: expected 8 fields, saw 9
Skipping line 220626: expected 8 fields, saw 9
Skipping line 227933: expected 8 fields, saw 11
Skipping line 228957: expected 8 fields, saw 10
Skipping line 245933: expected 8 fields, saw 9
Skipping line 251296: expected 8 fields, saw 9
Skipping line 259941: expected 8 fields, saw 9
Skipping line 261529: expected 8 fields, saw 9



# Data Understanding

## Book Dataset

In [41]:
book.head(2)

ISBN                title                author year release  \
0  0195153448  Classical Mythology    Mark P. O. Morford         2002   
1  0002005018         Clara Callan  Richard Bruce Wright         2001   

                 publisher                                            img_url  
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...  
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...

In [70]:
book.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   ISBN          271360 non-null  object
 1   title         271360 non-null  object
 2   author        271359 non-null  object
 3   year release  271360 non-null  object
 4   publisher     271358 non-null  object
 5   img_url       271357 non-null  object
dtypes: object(6)
memory usage: 12.4+ MB


In [35]:
book.shape

(271360, 8)

In [36]:
book.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [37]:
book = book[['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher', 'Image-URL-L']]

In [42]:
book.head(2)

ISBN                title                author year release  \
0  0195153448  Classical Mythology    Mark P. O. Morford         2002   
1  0002005018         Clara Callan  Richard Bruce Wright         2001   

                 publisher                                            img_url  
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...  
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...

In [39]:
book.rename(columns={
    'Book-Title' : 'title',
    'Book-Author' : 'author',
    'Year-Of-Publication' : 'year release',
    'Publisher' : 'publisher',
    'Image-URL-L' : 'img_url'
}, inplace=True)

### Missing Value Checking

In [71]:
book.isna().sum()

ISBN            0
title           0
author          1
year release    0
publisher       2
img_url         3
dtype: int64

In [73]:
print('Jumlah missing value :', book.shape[0] - book.dropna().shape[0])

Jumlah missing value : 6


**Observation**

1. Untuk Dataset buku terdapat 271360 baris dan 8 kolom, disini untuk nama kolom dirubah agar memudahkan pada saat analisis dan modeling
2. Terdapat missing value pada dataset buku yaitu sebanyak 6
2. Kolom ISBN merupakan data unik setiap buku

## User Dataset

In [48]:
user.head(2)

User-ID                   Location   Age
0        1         nyc, new york, usa   NaN
1        2  stockton, california, usa  18.0

In [74]:
user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   user_id   278858 non-null  int64  
 1   location  278858 non-null  object 
 2   age       168096 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB


In [49]:
user.shape

(278858, 3)

In [50]:
user.rename(columns={
    'User-ID' : 'user_id',
    'Location' :'location',
    'Age' : 'age'
}, inplace=True)

In [51]:
user.head(2)

user_id                   location   age
0        1         nyc, new york, usa   NaN
1        2  stockton, california, usa  18.0

### Missing Value Checking

In [75]:
user.isna().sum()

user_id          0
location         0
age         110762
dtype: int64

In [76]:
print('Jumlah missing value :', user.shape[0] - user.dropna().shape[0])

Jumlah missing value : 110762


**Observation**

1. Untuk Dataset user terdapat 278858 baris dan 3 kolom, disini untuk nama kolom dirubah agar memudahkan pada saat analisis dan modeling
2. Terdapat missing values sebanyak 110762
3. Kolom user_id merupakan data unik setiap user

## Rating Dataset

In [44]:
rating.head(2)

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5

In [77]:
rating.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 656605 entries, 412 to 1149637
Data columns (total 3 columns):
 #   Column   Non-Null Count   Dtype 
---  ------   --------------   ----- 
 0   user_id  656605 non-null  int64 
 1   ISBN     656605 non-null  object
 2   rating   656605 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 20.0+ MB


In [45]:
rating.shape

(1149780, 3)

In [46]:
rating.rename(columns={
    'User-ID' : 'user_id',
    'Book-Rating' :'rating'
}, inplace=True)

In [47]:
rating.head(2)

user_id        ISBN  rating
0   276725  034545104X       0
1   276726  0155061224       5

### Missing Value Checking

In [78]:
rating.isna().sum()

user_id    0
ISBN       0
rating     0
dtype: int64

In [79]:
print('Jumlah missing value :', rating.shape[0] - rating.dropna().shape[0])

Jumlah missing value : 0


**Observation**

1. Untuk Dataset rating terdapat 1149780 baris dan 3 kolom, disini untuk nama kolom dirubah agar memudahkan pada saat analisis dan modeling
2. Tidak terdapat missing value pada dataset rating
3. Terdapat kolom user_id dan ISBN yang merupakan data unik

## Total Data pada Masing-masing Dataset

In [52]:
print(f'Jumlah Data buku : {book.shape[0]}')
print(f'Jumlah Data user : {user.shape[0]}')
print(f'Jumlah Data rating : {rating.shape[0]}')

Jumlah Data buku : 271360
Jumlah Data user : 278858
Jumlah Data rating : 1149780


## Check Jumlah data pada dataset rating untuk user_id

In [56]:
rating['user_id'].value_counts()

11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
116180        1
116166        1
116154        1
116137        1
276723        1
Name: user_id, Length: 105283, dtype: int64

**Observation**

Dikarenakan banyaknya data pada rating dan terdapat user yang hanya memberikan 1 rating untuk setiap buku, hal ini akan mempengaruhi akurasi. Karena untuk recommender system Collaborative Filtering yaitu berdasarkan Behavior Interaction user atau interaksi user sehingga user yang memberikan rating semakin banyak semakin baik. Maka disini akan dilakukan filtering untuk mengambil user yang memberikan rating lebih dari 100

In [60]:
#filtering

cond = rating['user_id'].value_counts() > 100
cond[cond].shape

(1825,)

In [62]:
index = cond[cond].index
index

Int64Index([ 11676, 198711, 153662,  98391,  35859, 212898, 278418,  76352,
            110973, 235105,
            ...
            238186,  99441, 187410, 262070,  70183,  40553,  39345, 266283,
            189666, 140879],
           dtype='int64', length=1825)

**Observation**

Setelah dilakukan filtering hanya terdapat 1825 user yang memberikan rating diatas 100

In [67]:
#mengambil data yang hanya memberikan rating lebih dari 100
rating = rating[rating['user_id'].isin(index)]
rating.head(3)

user_id        ISBN  rating
412   276925  0006511929       0
413   276925  002542730X      10
414   276925  0060520507       0

In [69]:
print(f'Jumlah dataset rating setelah filtering : {rating.shape[0]}')

Jumlah dataset rating setelah filtering : 656605


## Merge Dataset and Feature Engineering 

### Merge Dataset rating_book dan book

In [80]:
rating_book = rating.merge(book, on='ISBN')
rating_book.head(3)

user_id        ISBN  rating  \
0   276925  002542730X      10   
1   277427  002542730X      10   
2     3363  002542730X       0   

                                               title             author  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   

  year release                  publisher  \
0         1994  John Wiley &amp; Sons Inc   
1         1994  John Wiley &amp; Sons Inc   
2         1994  John Wiley &amp; Sons Inc   

                                             img_url  
0  http://images.amazon.com/images/P/002542730X.0...  
1  http://images.amazon.com/images/P/002542730X.0...  
2  http://images.amazon.com/images/P/002542730X.0...

In [83]:
print(f'Jumlah data pada buku dan rating : {rating_book.shape[0]}')
print(f'Jumlah kolom pada buku dan rating : {rating_book.shape[1]}')

Jumlah data pada buku dan rating : 604854
Jumlah kolom pada buku dan rating : 8


**Observation**

Setelah dilakukan merge pada rating dengan book jumlah data menjadi 604854 dengan 8 kolom

### Melihat jumlah Rating pada setiap buku

In [87]:
num_rating = rating_book.groupby('title')['rating'].count().reset_index()
num_rating.head()

title  rating
0   A Light in the Storm: The Civil War Diary of ...       3
1                              Always Have Popsicles       1
2               Apple Magic (The Collector's series)       1
3   Beyond IBM: Leadership Marketing and Finance ...       1
4   Clifford Visita El Hospital (Clifford El Gran...       1

**Observation**

Hal ini dilakukan untuk melihat jumlah rating pada masing-masing buku. Terdapat buku yang memiliki rating lebih dari 1 yaitu buku A Light in the Storm 

In [89]:
#merename rating menjadi number of rating
num_rating.rename(columns={'rating':'num_of_rating'}, inplace=True)
num_rating.head(3)

title  num_of_rating
0   A Light in the Storm: The Civil War Diary of ...              3
1                              Always Have Popsicles              1
2               Apple Magic (The Collector's series)              1

### Merge Dataset rating_book dengan num_rating

In [90]:
#merge rating_book dengan num_rating agar terlihat total rating pada setiap buku

final_rating = rating_book.merge(num_rating, on='title')
final_rating.head(3)

user_id        ISBN  rating  \
0   276925  002542730X      10   
1   277427  002542730X      10   
2     3363  002542730X       0   

                                               title             author  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   

  year release                  publisher  \
0         1994  John Wiley &amp; Sons Inc   
1         1994  John Wiley &amp; Sons Inc   
2         1994  John Wiley &amp; Sons Inc   

                                             img_url  num_of_rating  
0  http://images.amazon.com/images/P/002542730X.0...            105  
1  http://images.amazon.com/images/P/002542730X.0...            105  
2  http://images.amazon.com/images/P/002542730X.0...            105

In [93]:
print(f'Jumlah data pada final_rating : {final_rating.shape[0]}')
print(f'Jumlah kolom pada final_rating : {final_rating.shape[1]}')

Jumlah data pada final_rating : 604854
Jumlah kolom pada final_rating : 9


**Observation**

Menggabungkan rating_book dengan num_rating sehingga Jumlah data menjadi 604854 dan kolom sebanyak 9

### Filtering Buku yang memiliki Jumlah Rating > 50

In [97]:
final_data = final_rating[final_rating['num_of_rating']>50]
final_data.head()

user_id        ISBN  rating  \
0   276925  002542730X      10   
1   277427  002542730X      10   
2     3363  002542730X       0   
3    10030  002542730X       7   
4    11676  002542730X       6   

                                               title             author  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
3  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
4  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   

  year release                  publisher  \
0         1994  John Wiley &amp; Sons Inc   
1         1994  John Wiley &amp; Sons Inc   
2         1994  John Wiley &amp; Sons Inc   
3         1994  John Wiley &amp; Sons Inc   
4         1994  John Wiley &amp; Sons Inc   

                                             img_url  num_of_rating  
0  http://images.amazon.com/images/P/002542730X.0...            105  
1  http://images.amazon.com/images/P/002542730X.0...            105  
2  http://images.amazon.com/images/P/002542730X.0...            105  
3  http://images.amazon.com/images/P/002542730X.0...            105  
4  http://images.amazon.com/images/P/002542730X.0...            105

In [100]:
print(f'Jumlah data pada final_data : {final_data.shape[0]}')
print(f'Jumlah kolom pada final_data : {final_data.shape[1]}')

Jumlah data pada final_data : 98726
Jumlah kolom pada final_data : 9


**Observation**

Dilakukan filtering buku yang memiliki jumlah rating lebih dari 50 karena pada data sebelumnya masih terdapat buku yang hanya memiliki jumlah rating 1 hal ini dapat mempengaruhi akurasi pada Recommender System.

Jumlah data menjadi 98726 dengan 9 kolom

# Data Cleaning

In [103]:
final_data.head(3)

user_id        ISBN  rating  \
0   276925  002542730X      10   
1   277427  002542730X      10   
2     3363  002542730X       0   

                                               title             author  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   

  year release                  publisher  \
0         1994  John Wiley &amp; Sons Inc   
1         1994  John Wiley &amp; Sons Inc   
2         1994  John Wiley &amp; Sons Inc   

                                             img_url  num_of_rating  
0  http://images.amazon.com/images/P/002542730X.0...            105  
1  http://images.amazon.com/images/P/002542730X.0...            105  
2  http://images.amazon.com/images/P/002542730X.0...            105

## Missing Value Handling

In [101]:
final_data.isna().sum()

user_id          0
ISBN             0
rating           0
title            0
author           0
year release     0
publisher        0
img_url          0
num_of_rating    0
dtype: int64

**Observation**

Tidak terdapat missing value pada final_data

## Duplicated Handling

In [105]:
final_data.duplicated(subset=['user_id','title']).sum()

2437

**Observation**

Terdapat duplicate value pada final_data sebanyak 2437 oleh karena itu akan dilakukan drop duplicate

In [106]:
print(f'Jumlah Data sebelum drop duplicated : {final_data.shape[0]}')
final_data = final_data.drop_duplicates(subset=['user_id','title'])
print(f'Jumlah Data setelah drop duplicated : {final_data.shape[0]}')

Jumlah Data sebelum drop duplicated : 98726
Jumlah Data setelah drop duplicated : 96289


# Pivot Table

In [109]:
book_pivot = final_data.pivot_table(columns='user_id', index='title', values='rating')
book_pivot.head()

user_id              183     254     507     882     1424    1435    1733    \
title                                                                         
1984                    NaN     9.0     NaN     NaN     NaN     NaN     NaN   
1st to Die: A Novel     NaN     NaN     0.0     NaN     NaN     NaN     NaN   
2010: Odyssey Two       NaN     NaN     NaN     NaN     NaN     NaN     NaN   
24 Hours                NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2nd Chance              NaN     NaN     NaN     NaN     NaN     NaN     NaN   

user_id              1903    2033    2110    ...  276018  276463  276680  \
title                                        ...                           
1984                    NaN     NaN     NaN  ...     NaN     NaN     NaN   
1st to Die: A Novel     NaN     NaN     NaN  ...     NaN     NaN     NaN   
2010: Odyssey Two       NaN     NaN     NaN  ...     NaN     NaN     NaN   
24 Hours                NaN     NaN     NaN  ...     NaN     NaN     NaN   
2nd Chance              NaN     NaN     NaN  ...     NaN     NaN     NaN   

user_id              276925  277427  277478  277639  278137  278188  278418  
title                                                                        
1984                    NaN     NaN     NaN     NaN     NaN     NaN     NaN  
1st to Die: A Novel     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
2010: Odyssey Two       NaN     NaN     NaN     NaN     NaN     NaN     NaN  
24 Hours                NaN    10.0     NaN     NaN     NaN     NaN     NaN  
2nd Chance              NaN     NaN     NaN     0.0     NaN     NaN     NaN  

[5 rows x 1793 columns]

In [110]:
book_pivot = book_pivot.fillna(0)
book_pivot.head()

user_id              183     254     507     882     1424    1435    1733    \
title                                                                         
1984                    0.0     9.0     0.0     0.0     0.0     0.0     0.0   
1st to Die: A Novel     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2010: Odyssey Two       0.0     0.0     0.0     0.0     0.0     0.0     0.0   
24 Hours                0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2nd Chance              0.0     0.0     0.0     0.0     0.0     0.0     0.0   

user_id              1903    2033    2110    ...  276018  276463  276680  \
title                                        ...                           
1984                    0.0     0.0     0.0  ...     0.0     0.0     0.0   
1st to Die: A Novel     0.0     0.0     0.0  ...     0.0     0.0     0.0   
2010: Odyssey Two       0.0     0.0     0.0  ...     0.0     0.0     0.0   
24 Hours                0.0     0.0     0.0  ...     0.0     0.0     0.0   
2nd Chance              0.0     0.0     0.0  ...     0.0     0.0     0.0   

user_id              276925  277427  277478  277639  278137  278188  278418  
title                                                                        
1984                    0.0     0.0     0.0     0.0     0.0     0.0     0.0  
1st to Die: A Novel     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
2010: Odyssey Two       0.0     0.0     0.0     0.0     0.0     0.0     0.0  
24 Hours                0.0    10.0     0.0     0.0     0.0     0.0     0.0  
2nd Chance              0.0     0.0     0.0     0.0     0.0     0.0     0.0  

[5 rows x 1793 columns]

**Observation**

1. Dilakukan pivot table untuk melihat user yang merating buku tersebut
2. mengisi nan value dengan nilai 0

In [111]:
from scipy.sparse import csc_matrix

In [113]:
book_sparse = csc_matrix(book_pivot)
book_sparse

<1060x1793 sparse matrix of type '<class 'numpy.float64'>'
	with 26439 stored elements in Compressed Sparse Column format>

# Modeling

## Train Model

In [116]:
from sklearn.neighbors import NearestNeighbors
model = NearestNeighbors(algorithm='brute')

In [117]:
model.fit(book_sparse)

NearestNeighbors(algorithm='brute')

In [153]:
distance, suggestion = model.kneighbors(book_pivot.iloc[329,:].values.reshape(1,-1), n_neighbors=6)

## Evaluation Model

In [154]:
distance

array([[ 0.        , 85.08231309, 86.48699324, 90.98901033, 98.79777325,
        99.05553998]])

In [155]:
suggestion

array([[329, 330, 332, 333, 847, 852]], dtype=int64)

In [156]:
for i in range(len(suggestion)):
    print(book_pivot.index[suggestion[i]])

Index(['Harry Potter and the Chamber of Secrets (Book 2)',
       'Harry Potter and the Goblet of Fire (Book 4)',
       'Harry Potter and the Prisoner of Azkaban (Book 3)',
       'Harry Potter and the Sorcerer's Stone (Book 1)',
       'The Klone and I: A High-Tech Love Story', 'The Laws of Our Fathers'],
      dtype='object', name='title')


In [159]:
#menyimpan book index pada variable book_name
book_name = book_pivot.index

In [160]:
import pickle
pickle.dump(model, open('books/model.pkl', 'wb'))
pickle.dump(book_name, open('books/book_name.pkl', 'wb'))
pickle.dump(final_data, open('books/final_data.pkl', 'wb'))
pickle.dump(book_pivot, open('books/book_pivot.pkl', 'wb'))


## Recommend Book Model

In [167]:
def recommend_book(book_name):
    book_id = np.where(book_pivot.index == book_name)[0][0]
    distance, suggestion = model.kneighbors(book_pivot.iloc[book_id,:].values.reshape(1,-1), n_neighbors=6)
    
    for i in range(len(suggestion)):
        books = book_pivot.index[suggestion[i]]
        for j in books:
            print(j)

In [170]:
book_name = 'Harry Potter and the Chamber of Secrets (Book 2)'
recommend_book(book_name)

Harry Potter and the Chamber of Secrets (Book 2)
Harry Potter and the Goblet of Fire (Book 4)
Harry Potter and the Prisoner of Azkaban (Book 3)
Harry Potter and the Sorcerer's Stone (Book 1)
The Klone and I: A High-Tech Love Story
The Laws of Our Fathers
